File: app/src/main/java/com/mdgd/pokemon/ui/pokemons/PokemonsViewModelFactory.java
package com.mdgd.pokemon.ui.pokemons;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.mdgd.pokemon.models.AppModule;

public class PokemonsViewModelFactory extends ViewModelProvider.NewInstanceFactory {
    private final AppModule appComponent;

    public PokemonsViewModelFactory(AppModule appComponent) {
        this.appComponent = appComponent;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass == PokemonsViewModel.class) {
            return (T) new PokemonsViewModel(appComponent.getPokemonsRepo());
        }
        return null;
    }
}

File: app/src/main/java/com/mdgd/pokemon/ui/pokemons/PokemonsViewModelFactory.kt
package com.mdgd.pokemon.ui.pokemons

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.mdgd.mvi.util.DispatchersHolderImpl
import com.mdgd.pokemon.models.AppModule

class PokemonsViewModelFactory(private val appComponent: AppModule) : ViewModelProvider.NewInstanceFactory() {

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return if (modelClass == PokemonsViewModel::class.java) {
            PokemonsViewModel(
                appComponent.getPokemonsRepo(),
                appComponent.getFiltersFactory(),
                DispatchersHolderImpl()
            ) as T
        } else super.create(modelClass)
    }
}
